File: rota-proxima/tests/test_commercial_agenda.py
import sys
import unittest
from pathlib import Path
from unittest.mock import patch

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
from dependency_stubs import install_optional_dependency_stubs
install_optional_dependency_stubs()
import server


def stop(id=1, owner='marcelo', date='2026-09-25', route_status='released'):
    return {'id': id, 'sequence': 1, 'status': 'pending', 'service_type': 'collection',
            'exact_time': '09:00:00', 'window_start': None, 'window_end': None,
            'pevs': {'id': id, 'name': 'PEV '+str(owner), 'city': 'Sorocaba', 'state': 'SP', 'commercial_owner_id': owner},
            'routes': {'id': id, 'name': 'Rota '+str(id), 'route_date': date, 'status': route_status},
            'internal_notes': 'não expor'}


class CommercialAgendaTests(unittest.TestCase):
    def handler(self, role='commercial', user_id='marcelo', query=None):
        h=server.AppHandler.__new__(server.AppHandler)
        h.require_user=lambda roles=None: {'id': user_id, 'role': role}
        h.token=lambda: 'user-token'
        h.query=lambda: query if query is not None else {'from':['2026-09-01'],'to':['2026-09-30']}
        h.send_json=lambda data,status=200,extra_headers=None:(status,data)
        return h

    def test_own_portfolio_even_when_another_commercial_is_requested(self):
        q={'from':['2026-09-01'],'to':['2026-09-30'],'commercial':['marcel'],'user_id':['marcel']}
        with patch.object(server.Supa,'get',return_value=[stop(),stop(2,'marcel')]) as get:
            status,data=self.handler(query=q).api_get('/api/agenda')
        self.assertEqual(200,status)
        self.assertEqual(['PEV marcelo'],[x['pev_name'] for x in data['items']])
        self.assertEqual('eq.marcelo',get.call_args.args[2]['pevs.commercial_owner_id'])
        self.assertEqual('user-token',get.call_args.args[1])

    def test_second_account_receives_only_its_pevs(self):
        with patch.object(server.Supa,'get',return_value=[stop(),stop(2,'marcel')]):
            _,data=self.handler(user_id='marcel').api_get('/api/agenda')
        self.assertEqual(['PEV marcel'],[x['pev_name'] for x in data['items']])

    def test_date_comes_from_route_without_depending_on_request_author(self):
        row=stop(date='2026-09-29')
        with patch.object(server.Supa,'get',return_value=[row]) as get:
            _,data=self.handler().api_get('/api/agenda')
        self.assertEqual('2026-09-29',data['items'][0]['route_date'])
        params=get.call_args.args[2]
        self.assertNotIn('requested_by',str(params))
        self.assertNotIn('scheduling_requests',params['select'])
        self.assertEqual('(route_date.gte.2026-09-01,route_date.lte.2026-09-30)',params['routes.and'])
        self.assertIn('routes!inner',params['select'])
        self.assertIn('pevs!inner',params['select'])

    def test_planned_released_active_and_finished_but_not_cancelled(self):
        rows=[stop(i+1,route_status=status) for i,status in enumerate(['draft','released','in_progress','finished','cancelled'])]
        with patch.object(server.Supa,'get',return_value=rows):
            _,data=self.handler().api_get('/api/agenda')
        self.assertEqual(['draft','released','in_progress','finished'],[x['route_status'] for x in data['items']])

    def test_outside_period_or_unowned_pevs_are_never_returned(self):
        rows=[stop(date='2026-10-01'),stop(2,owner=None),stop(3)]
        with patch.object(server.Supa,'get',return_value=rows):
            _,data=self.handler().api_get('/api/agenda')
        self.assertEqual([3],[x['id'] for x in data['items']])
        self.assertNotIn('internal_notes',data['items'][0])
        self.assertNotIn('commercial_owner_id',data['items'][0])

    def test_pagination_and_date_order(self):
        rows=[stop(i+1,date='2026-09-29') for i in range(500)]
        with patch.object(server.Supa,'get',side_effect=[rows,[stop(501,date='2026-09-10')]]) as get:
            _,data=self.handler().api_get('/api/agenda')
        self.assertEqual(501,len(data['items']))
        self.assertEqual(501,data['items'][0]['id'])
        self.assertEqual(['0','500'],[call.args[2]['offset'] for call in get.call_args_list])

    def test_invalid_periods_are_rejected_before_database_access(self):
        queries=[{}, {'from':['2026-09-25'],'to':['2026-09-01']},
                 {'from':['2026-02-30'],'to':['2026-09-30']},
                 {'from':['2025-01-01'],'to':['2026-09-30']},
                 {'from':['2026-9-1'],'to':['2026-09-30']}]
        with patch.object(server.Supa,'get') as get:
            for q in queries:
                with self.subTest(query=q):self.assertEqual(400,self.handler(query=q).api_get('/api/agenda')[0])
        get.assert_not_called()

    def test_other_roles_cannot_read_agenda(self):
        with patch.object(server.Supa,'get') as get:
            for role in ['admin','commercial_manager','quality','driver','production']:
                with self.subTest(role=role):self.assertEqual(403,self.handler(role).api_get('/api/agenda')[0])
        get.assert_not_called()

    def test_empty_portfolio(self):
        with patch.object(server.Supa,'get',return_value=[]):
            self.assertEqual((200,{'items':[]}),self.handler().api_get('/api/agenda'))

if __name__=='__main__':unittest.main()

File: rota-proxima/tests/test_commercial_agenda.mjs
import assert from 'node:assert/strict';
import fs from 'node:fs';
import test from 'node:test';
import vm from 'node:vm';
const source=fs.readFileSync(new URL('../static/app.js',import.meta.url),'utf8');
const section=source.slice(source.indexOf('function agendaStatus('),source.indexOf('async function renderHistory()'));
const row=(overrides={})=>({id:1,pev_name:'PEV Marcelo',city:'Sorocaba',state:'SP',route_name:'Rota 1',route_date:'2026-09-25',route_status:'released',status:'pending',service_type:'collection',...overrides});
function fixture({items=[row()],error=null,role='commercial'}={}) {
 const nodes=new Map(),calls=[],notices=[];
 const ctx=vm.createContext({
  state:{user:{role,name:'Marcelo'},page:'agenda'},
  $:selector=>{if(!nodes.has(selector))nodes.set(selector,{innerHTML:'',value:'',disabled:false,isConnected:true});return nodes.get(selector);},
  $$:()=>[],URLSearchParams,Intl,Date,
  esc:value=>String(value??'').replaceAll('&','&amp;').replaceAll('<','&lt;').replaceAll('>','&gt;').replaceAll('"','&quot;'),
  fmtDate:value=>value.split('-').reverse().join('/'),serviceTypeLabel:{collection:'Coleta',delivery:'Entrega'},
  api:async path=>{calls.push(path);if(error)throw error;return {items};},
  toast:(...args)=>notices.push(args),
 });
 vm.runInContext(section,ctx);
 return {ctx,nodes,calls,notices};
}
test('Agenda entra somente no menu Comercial',()=>{
 for(const role of ['commercial','admin','commercial_manager','driver','production','quality']){
  const {ctx,nodes}=fixture({role});
  vm.runInContext(source.slice(source.indexOf('function renderNav()'),source.indexOf('async function go(page)')),ctx);
  ctx.renderNav();assert.equal(nodes.get('#nav').innerHTML.includes('data-page="agenda"'),role==='commercial');
 }
});
test('agrupa por data programada em ordem cronológica',()=>{
 const {ctx}=fixture();
 const html=ctx.agendaListHtml([row(),row({id:2,route_date:'2026-09-24',pev_name:'PEV anterior'}),row({id:3})]);
 assert.equal((html.match(/<section/g)||[]).length,2);
 assert.ok(html.indexOf('24/09/2026')<html.indexOf('25/09/2026'));
 assert.match(html,/2 agendamentos/);
});
test('situações de agendamento e execução têm nomes distintos',()=>{
 const {ctx}=fixture();
 for(const [input,label] of [[{route_status:'draft'},'Planejada'],[{},'Agendada'],[{route_status:'in_progress'},'Em rota'],[{status:'arrived'},'No local'],[{status:'completed'},'Realizada'],[{status:'failed'},'Não realizada']]){
  assert.equal(ctx.agendaStatus(row(input))[1],label);
 }
});
test('horário exato, janela e ausência de horário',()=>{
 const {ctx}=fixture();
 assert.match(ctx.agendaListHtml([row({exact_time:'09:00:00'})]),/Horário previsto: 09:00/);
 assert.match(ctx.agendaListHtml([row({window_start:'08:00:00',window_end:'12:00:00'})]),/Janela: 08:00 às 12:00/);
 assert.match(ctx.agendaListHtml([row()]),/Horário não definido/);
});
test('pesquisa e escapes impedem HTML em nomes',()=>{
 const {ctx}=fixture();
 const rows=[row({pev_name:'<script>teste</script>'}),row({id:2,pev_name:'Outro PEV',city:'Votorantim'})];
 const html=ctx.agendaListHtml(rows,'TESTE');
 assert.match(html,/&lt;script&gt;/);assert.doesNotMatch(html,/<script>|Outro PEV/);
 assert.match(ctx.agendaListHtml(rows,'Votorantim'),/Outro PEV/);
 assert.match(ctx.agendaListHtml(rows,'inexistente'),/Nenhum agendamento/);
});
test('consulta envia apenas datas e usa o usuário da sessão no servidor',async()=>{
 const {ctx,nodes,calls}=fixture();
 // O DOM de teste não interpreta o value do HTML; preparar entradas equivalentes.
 ctx.$('#agendaFrom').value='2026-09-01';ctx.$('#agendaTo').value='2026-09-30';
 await ctx.renderAgenda();
 assert.equal(calls[0],'/api/agenda?from=2026-09-01&to=2026-09-30');
 assert.match(nodes.get('#page').innerHTML,/Comercial • Marcelo/);
 assert.match(nodes.get('#agendaResults').innerHTML,/PEV Marcelo/);
 assert.equal(nodes.get('#agendaRefresh').disabled,false);
});
test('falha de consulta limpa resultado e permite tentar novamente',async()=>{
 const {ctx,nodes}=fixture({error:new Error('Dados indisponíveis')});
 ctx.$('#agendaFrom').value='2026-09-01';ctx.$('#agendaTo').value='2026-09-30';
 await ctx.renderAgenda();
 assert.match(nodes.get('#agendaResults').innerHTML,/role="alert".*Dados indisponíveis/);
 assert.equal(nodes.get('#agendaRefresh').disabled,false);
});
test('perfil sem permissão não dispara consulta',async()=>{
 const {ctx,calls}=fixture({role:'quality'});
 await assert.rejects(ctx.renderAgenda(),/Sem permissão/);assert.equal(calls.length,0);
});
